File: item_price/public/js/item_price.js
frappe.ui.form.on('Item Price', {
    custom_price_type:function(frm){
        if(frm.doc.custom_price_type == 'Last Purchase Rate'){
            frappe.call({
                method:'item_price.public.py.item_price.get_last_purchase_rate',
                args:{'item_code':frm.doc.item_code}
            }).then(records =>{
                var data =  records['data']
                frm.set_value('price_list_rate', data)
            })
        }
        else if (frm.doc.custom_price_type == 'Item Average'){
            frappe.call({
                method:'item_price.public.py.item_price.get_avg_price',
                args:{'item_code':frm.doc.item_code, 'company':frm.doc.custom_company}
            }).then(records =>{
                var data =  records['data']

                length = data.length
                
                avr_rate = 0
                data.forEach(function(obj){
                     avr_rate += obj['valuation_rate']
                })

                finial_ave = avr_rate / length
                frm.set_value('price_list_rate', finial_ave)
            })
        }
    }
})

File: item_price/public/py/item_price.py
import frappe

@frappe.whitelist()
def get_last_purchase_rate(item_code):
    
    data = frappe.db.sql("""
    SELECT rate, qty, landed_cost_voucher_amount
    FROM `tabPurchase Receipt Item`
    WHERE item_code = %s
    ORDER BY creation DESC
    LIMIT 1
""", (item_code,), as_dict=True)

    last_rate = 0
    for i in data:
        last_landed = i['landed_cost_voucher_amount'] / i['qty']
        last_rate = last_landed + i['rate']

    frappe.response['data']=last_rate


@frappe.whitelist()
def get_avg_price(item_code, company):

    data = frappe.db.sql("""
    SELECT b.valuation_rate
    FROM `tabBin` b
    INNER JOIN `tabWarehouse` w ON b.warehouse = w.name
    WHERE b.item_code = %s AND w.company = %s
""", (item_code, company), as_dict=True)   


    frappe.response['data']=data
    